src/Numerics/ODESolvers/GenericCallbacks.jl

```diff
@@ -1,137 +1,193 @@
 """
     GenericCallbacks
 
-A set of callback functions to be used with an `AbstractODESolver`
+This module defines interfaces and wrappers for callbacks to be used with an
+`AbstractODESolver`.
+
+A callback `cb` defines two methods:
+
+- `GenericCallbacks.init!(cb, solver, Q, param, t0)`, to be called at solver
+  initialization.
+
+- `GenericCallbacks.call!(cb, solver, Q, param, t)`, to be called after each time step:
+  the return value dictates what action should be taken:
+
+   * `0` or `nothing`: continue time stepping as usual
+   * `1`: stop time stepping after all callbacks have been executed
+   * `2`: stop time stepping immediately
+
+Additionally, "wrapper" callbacks can be used to execute the callbacks under certain
+conditions:
+
+ - [`AtStart`](@ref)
+ - [`EveryXWallTimeSeconds`](@ref)
+ - [`EveryXSimulationTime`](@ref)
+ - [`EveryXSimulationSteps`](@ref)
+
+For convenience, the following objects can also be used as callbacks:
+
+ - a `Function` object `f`, `init!` is a no-op, and `call!` will call `f()`, and ignore the return value.
+ - a `Tuple` object will call `init!` and `call!` on each element of the tuple.
+
 """
 module GenericCallbacks
+
+export AtStart,
+    EveryXWallTimeSeconds, EveryXSimulationTime, EveryXSimulationSteps
+
 using MPI
 
-using ..ODESolvers
+init!(f::Function, solver, Q, param, t0) = nothing
+function call!(f::Function, solver, Q, param, t0)
+    f()
+    return nothing
+end
+
+function init!(callbacks::Tuple, solver, Q, param, t0)
+    for cb in callbacks
+        GenericCallbacks.init!(cb, solver, Q, param, t0)
+    end
+end
+
+function call!(callbacks::Tuple, solver, Q, param, t)
+    val = 0
+    for cb in callbacks
+        val_i = GenericCallbacks.call!(cb, solver, Q, param, t)
+        val_i = (val_i === nothing) ? 0 : val_i
+        val = max(val, val_i)
+        if val == 2
+            return val
+        end
+    end
+    return val
+end
+
+
+abstract type AbstractCallback end
 
 """
-    EveryXWallTimeSeconds(f, time, mpicomm)
+    AtStart(callback) <: AbstractCallback
 
-This callback will run the function 'f()' every `time` wallclock time seconds.
-The `mpicomm` is used to syncronize runtime across MPI ranks.
+A wrapper callback to execute `callback` at initialization, in addition to after each
+timestep.
 """
-struct EveryXWallTimeSeconds
-    "time of the last callback"
-    lastcbtime_ns::Array{UInt64}
-    "time between callbacks"
+struct AtStart <: AbstractCallback
+    callback
+end
+function init!(cb::AtStart, solver, Q, param, t0)
+    init!(cb.callback, solver, Q, param, t0)
+    call!(cb.callback, solver, Q, param, t0)
+end
+
+function call!(cb::AtStart, solver, Q, param, t)
+    call!(cb.callback, solver, Q, param, t)
+end
+
+
+
+
+
```

Comment:
```suggestion
```
Does JuliaFormatter take care of these additional blank lines?

Reply:
I think there's an option for it